tests/v3_api/test_auth.py

```diff
@@ -0,0 +1,333 @@
+        access_mode, add_users_to_cluster=True):
+    delete_cluster_users()
+    client = get_admin_client()
+    for project in client.list_project():
+        delete_existing_users_in_project(client, project)
+    auth_setup_data = setup["auth_setup_data"]
+    admin_user = auth_setup_data["admin_user"]
+    token = login(admin_user, PASSWORD)
+    allowed_principal_ids = []
+    allowed_principal_ids.append(principal_lookup(admin_user, token))
+
+    # Add users and groups in allowed list to access rancher-server
+    add_users_to_siteAccess(token, access_mode, allowed_principal_ids)
+
+    groups_to_check = []
+    users_to_check = []
+    if add_users_to_cluster:
+        groups_to_check = auth_setup_data["groups_added_to_cluster"]
+        users_to_check = auth_setup_data["users_added_to_cluster"]
+    else:
+        groups_to_check = auth_setup_data["groups_added_to_project"]
+        users_to_check = auth_setup_data["users_added_to_project"]
+    for group in groups_to_check:
+        for user in auth_setup_data[group]:
+            login(user, PASSWORD,
+                  expected_status=setup["permission_denied_code"])
+
+    for user in users_to_check:
+        login(user, PASSWORD,
+              expected_status=setup["permission_denied_code"])
+
+    client = get_client_for_token(token)
+    for group in groups_to_check:
+        if add_users_to_cluster:
+            assign_user_to_cluster(client, principal_lookup(group, token),
+                                   setup["cluster1"], "cluster-owner")
+        else:
+            assign_user_to_project(client, principal_lookup(group, token),
+                                   setup["project2"], "project-owner")
+    for user in users_to_check:
+        if add_users_to_cluster:
+            assign_user_to_cluster(client, principal_lookup(user, token),
+                                   setup["cluster1"], "cluster-owner")
+        else:
+            assign_user_to_project(client, principal_lookup(user, token),
+                                   setup["project2"], "cluster-owner")
+    expected_status = setup["permission_denied_code"]
+
+    if access_mode == "required":
+        expected_status = setup["permission_denied_code"]
+
+    if access_mode == "restricted":
+        expected_status = 201
+
+    for group in groups_to_check:
+        for user in auth_setup_data[group]:
+            login(user, PASSWORD, expected_status)
+
+    for user in users_to_check:
+        login(user, PASSWORD, expected_status)
+
+
+def login(username, password, expected_status=201):
+    token = ""
+    r = requests.post(CATTLE_AUTH_URL, json={
+        'username': username,
+        'password': password,
+        'responseType': 'json',
+    }, verify=False)
+    assert r.status_code == expected_status
+    print "Login request for " + username + " " + str(expected_status)
+    if expected_status == 201:
+        token = r.json()['token']
+    return token
+
```

Comment:
Can we rearrange the functions in the file such that configuration/login functions are first and then the testcases that depend on these steps - this will make it easier for readability.